src/stores/issue-page.tsx

```diff
@@ -0,0 +1,201 @@
+    case 'CHANGE_STATUS_INITIATE': {
+      const { issueState } = initialState
+      if (issueState.loading) {
+        throw new Error('Posting comments while the issue is loading should not be possible')
+      }
+      const { issue } = issueState
+
+      if (!issue) {
+        throw new Error('Posting comments for a nonexistent issue should not be possible')
+      }
+
+      if (initialState.actionInProgress) {
+        throw new Error('Wait for the current action to finish before starting another')
+      }
+
+      const now = new Date()
+
+      const nextTimeline: IssueTimeline = issue.timeline.concat([
+        {
+          verb: 'comment',
+          by: action.payload.user,
+          timestamp: now,
+          comment: action.payload.comment,
+        },
+        {
+          verb: 'change status',
+          by: action.payload.user,
+          timestamp: now,
+          status: action.payload.status,
+        },
+      ])
+
+      const nextIssue: Issue = {
+        ...issue,
+        aggregates: {
+          ...issue.aggregates,
+          comments: {
+            count: issue.aggregates.comments.count + 1,
+          },
+        },
+        timeline: nextTimeline,
+      }
+
+      return {
+        ...initialState,
+        issueState: { loading: false, issue: nextIssue },
+        actionInProgress: { priorState: initialState, action },
+      }
+    }
+
+    // Optimistically update the issue with the timeline that would result from the action.
+    // Set the actionInProgress with the prior state and the action so that this can be rolled back on an error.
+    case 'POST_COMMENT_INITIATE': {
+      const { issueState } = initialState
+      if (issueState.loading) {
+        throw new Error('Posting comments while the issue is loading should not be possible')
+      }
+      const { issue } = issueState
+
+      if (!issue) {
+        throw new Error('Posting comments for a nonexistent issue should not be possible')
+      }
+
+      if (initialState.actionInProgress) {
+        throw new Error('Wait for the current action to finish before starting another')
+      }
+
+      const nextTimeline: IssueTimeline = issue.timeline.concat([
+        {
+          verb: 'comment',
+          by: action.payload.user,
+          timestamp: new Date(),
+          comment: action.payload.comment,
+        },
+      ])
+
+      const nextIssue: Issue = {
+        ...issue,
+        aggregates: {
+          ...issue.aggregates,
+          comments: {
+            count: issue.aggregates.comments.count + 1,
+          },
+        },
+        timeline: nextTimeline,
+      }
+
+      return {
+        ...initialState,
+        issueState: { loading: false, issue: nextIssue },
+        actionInProgress: { priorState: initialState, action },
+      }
+    }
+
+    // The successful completion of the action means that the optimistic update can stay and there's no action in progress.
+    case 'CHANGE_STATUS_SUCCESS':
```

Comment:
minor - suggest combining this case with POST_COMMENT_SUCCESS (similar to CHANGE_STATUS_ERROR/POST_COMMENT_ERROR)